### ISSUE TITLE
Upgrade from previous version fails

### ISSUE THREAD

**Issue by lat9** [lat9/usu#19]
As provided on [this ](https://www.zen-cart.com/showthread.php?199064-Ultimate-SEO-2-200-Support-Thread&p=1355901#post1355901)Zen Cart support-thread post (thanks, @pan2020):
![image](https://user-images.githubusercontent.com/2685585/54316397-0400be00-45b7-11e9-8850-948c92aa2c03.png)
... the upgrade processing provided by USU 3.0.0+ is not detecting those left-over settings from a previous (v2x or SEO URLs) installation.

**Comment by lat9** [comment]
There's now a fairly comprehensive restructuring of the plugin's installation/upgrade processing.  Essentially, once v3.0.1 (or later) is installed, the next logged-in admin page's processing will (re-)install USU for configured versions less than 3.0.1.

If an `Ultimate SEO` configuration group is found, it will be used to house the configuration settings.  Otherwise, if an `Ultimate URLs` group is found, it will be used; otherwise, a new configuration group will be created.

On an initial install (i.e. no `USU_VERSION` previously defined) or on a version-change from a USU version _prior to_ 3.1.0, the processing will search for any `Ultimate SEO` settings, convert them to current USU values and use those as the default values.  All `Ultimate SEO` settings are removed from the database as part of the install process.

Next, a check is made to see if any of the current `USU_*` settings are defined, using those as the default values (and overriding any previous `Ultimate SEO` settings, in case both are present).  All pre-existing USU settings are removed from the database and all configuration settings are re-installed.

**Comment by Pan2020** [comment]
So I downloaded your latest files from Github and replaced all PHP files into the server.

As soon as I logged into admin, this notice showed up (ignore the yellow one, since it is expected due to admin log from production server...)

![image](https://user-images.githubusercontent.com/21959676/55834163-3e427a00-5ae7-11e9-8117-46924a986c51.png)

The settings are all seems to be back to default values.

![image](https://user-images.githubusercontent.com/21959676/55834322-a3966b00-5ae7-11e9-8d5b-96ab1ace85ea.png)

The database were also erased the SEO variant settings, while overrode the USU settings to latest dates.

![image](https://user-images.githubusercontent.com/21959676/55834396-ccb6fb80-5ae7-11e9-829d-1cd74f306938.png)

I think if this is your intend... You should tell in the documentation that settings are going to be revert back to defaults and suggest users to write down settings BEFORE installing/upgrading USU to 3.0.1 (or later)!?

**Comment by lat9** [comment]
The upgrade process **_should have_** kept your old settings.  If you still have a copy of the database that you used on the upgrade, would you zip-up the contents of the `configuration` table for those pre-upgrade Ultimate URLs settings and post that back here?

**Comment by lat9** [comment]
Hmm, comparing the original phpMyAdmin screenshot to the most recent posted, it looks like those settings _were_ carried over from the previous version.  The upgrade processing:
1. Checks to see if there are any 'Ultimate SEO'-related settings in the database and, if so, uses them as the default value(s).  The 'Ultimate SEO' settings are then removed from the database.
2. Checks to see if there are any preexisting 'Ultimate URLs' settings and, if so, uses them as the default values (other than the version!).  Those 'Ultimate URLs' settings are then removed from the database, as their `set_function` and `use_function` settings have all changed.
3. Inserts the updated 'Ultimate URLs' settings, with values possibly overridden, into the database.

So, upon completion of the USU 3.0.1 upgrade, all of the 'Ultimate URLs' settings will, as you indicated, now have a `date_added` of **_now()_**.

Note that, in the case posted, if both `SEO_` and `USU_` settings are present in the database, the value set for the associated `USU_` one is used.

**Comment by lat9** [comment]
In reviewing that original posting, I saw that I'd missed the carry-over of a couple of the Ultimate SEO settings; they're now included!

Note also that the install/upgrade-to-v301-or-later processing now creates a `/logs/usu-yyyymmdd-hhmmss.log` file that records its gathering of previous settings at each step.

**Comment by Pan2020** [comment]
[usu-install-20190409-203716.log](https://github.com/lat9/usu/files/3070101/usu-install-20190409-203716.log)
[usu-install-20190410-210538.log](https://github.com/lat9/usu/files/3070102/usu-install-20190410-210538.log)
[usu-install-20190411-163853.log](https://github.com/lat9/usu/files/3070100/usu-install-20190411-163853.log)

Here are install logs by USU. I did couple reinstall as I try to finish up the testing on ZC 1.5.6 and upgrade the ZC version on production site.
It seems that USU plugin is surely fixed on most (if not all) stuffs!

**Comment by Pan2020** [comment]
Update!

A bit of background...

As I'm on the late-stages of testing ZC 1.5.6a/b (I keep update files directly to Github to test it out!)
I'm also testing essential add-ons to ensure that new ZC site have most functions that are currently available in ZC 1.5.1!
(Some had to use substitute add-on due to add-on currently in use for ZC 1.5.1 were either no longer supported, no update for years, merged with ZC core, or became another plugin used for newer ZC.)

I had Super Order, Edit Order, and COWOA from DivaVocal and others... However, DivaVocal seems to be moved on Zen Cart (I don't see she comes to ZC site as much anymore...) and those add on were taken over or merged or became another add-on instead. Hence, I had to use USU to substitute the COWOA functions and I cannot put ZC 1.5.6 into production mode until the add-on I needed are properly functional!

Right now, I have ZC 1.5.6 installed and reinstalled each time there is an update.
So, after I reinstalled... bunch of green stacks were stacked up!
https://i.imgur.com/PebfnRH.png

However, the IH notice still stay, as I stated in the IH topic at ZC forum, I will continue on that topic there... https://www.zen-cart.com/showthread.php?222983-Image-Handler-5-(for-v1-5-5)-Support-Thread&p=1356694#post1356694

Back to USU, so it seems settings and versions were updated.
However... I compared to two databases, I do see something...

ZC151 DB: https://i.imgur.com/Qls0u3T.png
ZC156 DB: https://i.imgur.com/bqBxU1b.png

So like `ENABLED` setting is now `false` (is this intended due to testing?)
The `CPATH` from `off` to `auto`
The `CATEGORY_DIR` from `full` to `short`

Basically, settings are still not kept from the original database with `SEO` variant settings!

So it seems, no settings were converted! I do not know if parameters were even pass through!
Since based on the log... It did not even dumb the add-on versions!
(Confirmed with Notepad++ search by copied whole line of dumped values... all 4 were same values, that was FALSE!)
https://i.imgur.com/6h1FthW.png

So it seems either database did not pick up values properly or somehow the parameter just missed the function arguments entirely!

**Comment by lat9** [comment]
Well, crud.  I was not using the correct variable name for the upgrade-from-seo-urls' processing.  I've used your configuration settings (thanks a bunch!) to validate that the `Ultimate SEO` settings are properly carried over.  Give that update another try.
